ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/InsertValuesExecutor.java

```diff
@@ -0,0 +1,244 @@
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server.execution;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.KsqlExecutionContext;
+import io.confluent.ksql.logging.processing.ProcessingLogContext;
+import io.confluent.ksql.metastore.model.KsqlStream;
+import io.confluent.ksql.metastore.model.StructuredDataSource;
+import io.confluent.ksql.parser.tree.AstVisitor;
+import io.confluent.ksql.parser.tree.Expression;
+import io.confluent.ksql.parser.tree.InsertValues;
+import io.confluent.ksql.parser.tree.Literal;
+import io.confluent.ksql.parser.tree.NullLiteral;
+import io.confluent.ksql.rest.entity.KsqlEntity;
+import io.confluent.ksql.services.ServiceContext;
+import io.confluent.ksql.statement.ConfiguredStatement;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.SchemaUtil;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.LongSupplier;
+import java.util.stream.Collectors;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.connect.data.Field;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.Schema.Type;
+
+public class InsertValuesExecutor {
+
+  private static final long MAX_SEND_TIMEOUT_SECONDS = 5;
+
+  private final LongSupplier clock;
+
+  public InsertValuesExecutor() {
+    this(System::currentTimeMillis);
+  }
+
+  @VisibleForTesting
+  InsertValuesExecutor(final LongSupplier clock) {
+    this.clock = Objects.requireNonNull(clock, "clock");
+  }
+
+  public Optional<KsqlEntity> execute(
+      final ConfiguredStatement<InsertValues> statement,
+      final KsqlExecutionContext executionContext,
+      final ServiceContext serviceContext
+  ) {
+    final InsertValues insertValues = statement.getStatement();
+    final StructuredDataSource<?> dataSource = executionContext
+        .getMetaStore()
+        .getSource(insertValues.getTarget().getSuffix());
+
+    if (dataSource instanceof KsqlStream && ((KsqlStream<?>) dataSource).hasWindowedKey()) {
+      throw new KsqlException("INSERT VALUES does not support windowed data sources!");
```

Comment:
Data source is a bit ambiguous here since it's indeed sink for the statement, maybe change the error message to something like `You cannot insert values into stream/table with windowed key`!